File: shared/src/commonMain/kotlin/com.commuto.interfacemobile.dbService/DBService.kt
package com.commuto.interfacemobile.dbService

import com.commuto.interfacemobile.db.Database
import com.commuto.interfacemobile.db.DatabaseDriverFactory
import com.commuto.interfacemobile.db.KeyPair
import com.commuto.interfacemobile.db.PublicKey

/**
 * The Database Service Class.
 *
 * This is responsible for storing data, serving it to Local Services upon request, and accepting
 * Local Services' requests to add and remove data from storage.
 *
 * @property databaseDriverFactory the DatabaseDriverFactory used to obtain a platform-specific
 * driver that DBService will use to interact with the database.
 */
class DBService(databaseDriverFactory: DatabaseDriverFactory) {
    private val database = Database(databaseDriverFactory)
    //TODO: Localize error strings
    /**
     * Creates all necessary tables for proper operation.
     * No-op on mobile
     */
    fun createTables() {
    }

    /**
     * Clears entire database
     */
    fun clearDatabase() {
        database.clearDatabase()
    }

    /**
     * Persistently stores a key pair associated with an interface id.
     *
     * @param interfaceId the interface id of the keypair as a byte array encoded to a hexadecimal String
     * @param publicKey the public key of the keypair as a byte array encoded to a hexadecimal String
     * @param privateKey the private key of the keypair as a byte array encoded to a hexadecimal String
     *
     * Used exclusively by KMService
     */
    fun storeKeyPair(interfaceId: String, publicKey: String, privateKey: String) {
        val keyPair = KeyPair(interfaceId, publicKey, privateKey)
        if (getKeyPair(interfaceId) != null) {
            throw IllegalStateException("Database query for key pair with interface id " + interfaceId +
                    " returned result")
        } else {
            database.insertKeyPair(keyPair)
        }
    }

    /**
     * Retrieves the persistently stored key pair associated with the given interface id, or returns
     * null if not present.
     *
     * @param interfaceId the interface id of the keypair as a byte array encoded to a hexadecimal String
     *
     * @return KeyPair object (defined in com.commuto.interfacedesktop.db) if key pair is found
     * @return null if key pair is not found
     * @throws IllegalStateException if multiple key pairs are found for a given interface id, or if the interface id
     * returned from the database query does not match the interface id passed to getKeyPair
     *
     * Used exclusively by KMService
     */
    fun getKeyPair(interfaceId: String): KeyPair? {
        val dbKeyPairs: List<KeyPair> = database.selectKeyPairByInterfaceId(interfaceId)
        if (dbKeyPairs.size > 1) {
            throw IllegalStateException("Multiple key pairs found with given interface id " + interfaceId)
        } else if (dbKeyPairs.size == 1) {
            check(dbKeyPairs[0].interfaceId.equals(interfaceId)) {
                "Returned interface id " + dbKeyPairs[0].interfaceId + "did not match specified interface id " +
                        interfaceId
            }
            return KeyPair(interfaceId, dbKeyPairs[0].publicKey, dbKeyPairs[0].privateKey)
        } else {
            return null
        }
    }

    /**
     * Persistently stores a public key associated with an interface id.
     *
     * @param interfaceId the interface id of the keypair as a byte array encoded to a hexadecimal String
     * @param publicKey the public key to be stored as a byte array encoded to a hexadecimal String
     *
     * @throws IllegalStateException if a public key with the given interface id is already found in the database
     */
    fun storePublicKey(interfaceId: String, publicKey: String) {
        val publicKey = PublicKey(interfaceId, publicKey)
        if (getPublicKey(interfaceId) != null) {
            throw IllegalStateException("Database query for public key with interface id " + interfaceId +
                    " returned result")
        } else {
            database.insertPublicKey(publicKey)
        }
    }

    /**
     * Retrieves the persistently stored public key associated with the given interface id, or returns
     * null if not present
     *
     * @param interfaceId the interface id of the public key as a byte array encoded to a hexadecimal String
     *
     * @return PublicKey object (definded in com.commuto.interfacedesktop.db) if public key is found
     * @return null if public key is not found
     * @throws IllegalStateException if multiple public keys are found for a given interface id, or if the interface id
     * returned from the database query does not match the interface id passed to getPublicKey
     *
     * Used exclusively by KMService
     */
    fun getPublicKey(interfaceId: String): PublicKey? {
        val dbPublicKeys: List<PublicKey> = database.selectPublicKeyByInterfaceId(interfaceId)
        if (dbPublicKeys.size > 1) {
            throw IllegalStateException("Multiple public keys found with given interface id " + interfaceId)
        } else if (dbPublicKeys.size == 1) {
            check(dbPublicKeys[0].interfaceId.equals(interfaceId)) {
                "Returned interface id " + dbPublicKeys[0].interfaceId + "did not match specified interface id " +
                        interfaceId
            }
            return PublicKey(interfaceId, dbPublicKeys[0].publicKey)
        } else {
            return null
        }
    }
}

File: iosApp/iosApp/dbService/DBService.swift
//
//  DBService.swift
//  iosApp
//
//  Created by jimmyt on 11/27/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import SQLite

#warning("TODO: rename this as DatabaseService")
/**
 The Database Service Class.
 
 This is responsible for storing data, serving it to other services upon request, and accepting services' requests to add and remove data from storage.
 */
class DBService {
    
    #warning("TODO: rename this as connection")
    /**
     The connection to the SQLite database.
     */
    var db: Connection?
    
    /**
     The serial DispatchQueue used to synchronize database access. All database queries should be run on this queue to prevent data races.
     */
    let databaseQueue = DispatchQueue(label: "databaseService.serial.queue")
    
    /**
     A database table of [OfferOpened](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#offeropened) events.
     */
    let offerOpenedEvents = Table("OfferOpenedEvent")
    /**
     A database table of `KeyPair`s.
     */
    let keyPairs = Table("KeyPair")
    /**
     A database table of `PublicKey`s.
     */
    let publicKeys = Table("PublicKey")
    
    /**
     A database structure representing an offer ID.
     */
    let offerId = Expression<String>("offerId")
    /**
     A database structure representing an interface ID.
     */
    let interfaceId = Expression<String>("interfaceId")
    /**
     A database structure representing a public key.
     */
    let publicKey = Expression<String>("publicKey")
    /**
     A database structure representing a private key.
     */
    let privateKey = Expression<String>("privateKey")
    
    #warning("This should get its own file")
    /**
     An `Error` thrown by `DatabaseService` functions.
     */
    enum DBServiceError: Error {
        /**
         Thrown when `DatabaseService` receives an unexpected result from a database query.
         
         - Parameter message: A `String` that provides information about the context in which the error was thrown.
         */
        case unexpectedDbResult(message: String)
    }
    
    /**
     Creates all necessary database tables.
     */
    func createTables() throws {
        try db!.run(keyPairs.create { t in
            t.column(interfaceId, unique: true)
            t.column(publicKey)
            t.column(privateKey)
        })
        try db!.run(publicKeys.create { t in
            t.column(interfaceId, unique: true)
            t.column(publicKey)
        })
        try db!.run(offerOpenedEvents.create { t in
            t.column(offerId, unique: true)
            t.column(interfaceId)
        })
    }
    
    /**
     Creates a new in-memory database, and stores its reference in `DatabaseService`'s `db` property.
     */
    func connectToDb() throws {
        db = try Connection(.inMemory)
    }
    
    /**
     Persistently stores the contents of an [OfferOpened](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#offeropened) Event.
     
     - Parameters:
        - id: The offer ID specified in the OfferOpened event, as a `String`.
        - interfaceId: The interface ID specified in the OfferOpened event, as a `String`.
     */
    func storeOfferOpenedEvent(id: String, interfaceId: String) throws {
        _ = try databaseQueue.sync {
            try db!.run(offerOpenedEvents.insert(offerId <- id, self.interfaceId <- interfaceId))
        }
    }
    
    /**
     Persistently stores a key pair associated with an interface ID.
     
     - Note: This function is used exclusively by `KeyManagerService`
     
     - Parameters:
        - interfaceId: The interface ID of the key pair as a byte array encoded to a hexadecimal `String`.
        - publicKey: The public key of the key pair as a byte array encoded to a hexadecimal `String`.
        - privateKey: The private key of the key pair as a byte array encoded to a hexadecimal `String`.
     */
    func storeKeyPair(interfaceId interf_id: String, publicKey pub_key: String, privateKey priv_key: String) throws {
        guard try getKeyPair(interfaceId: interf_id) == nil else {
            throw DBServiceError.unexpectedDbResult(message: "Database query for key pair with interface id " + interf_id + " returned result")
        }
        try db!.run(keyPairs.insert(interfaceId <- interf_id, publicKey <- pub_key, privateKey <- priv_key))
    }
    
    /**
     Retrieves the persistently stored key pair associated with the given interface ID, or returns `nil` if no such key pair is present.
     
     - Parameter interfaceId: The interface ID of the desired key pair as a byte array encoded to a hexadecimal `String`.
     
     - Returns: A `DBKeyPair` if a key pair with the specified interface ID is found, or `nil` if such a key pair is not found.
     
     - Throws: A `DBServiceError.inexpectedDbResult` if multiple key pairs are found for a single interface ID, or if the interface ID of the key pair returned from the database query does not match `interfaceId`.
     */
    func getKeyPair(interfaceId interfId: String) throws -> DBKeyPair? {
        let rowIterator = try db!.prepareRowIterator(keyPairs.filter(interfaceId == interfId))
        let result = try Array(rowIterator)
        if (result.count > 1) {
            throw DBServiceError.unexpectedDbResult(message: "Multiple key pairs found with given interface id " + interfId)
        } else if (result.count == 1) {
            guard result[0][interfaceId] == interfId else {
                throw DBServiceError.unexpectedDbResult(message: "Returned interface id " + result[0][interfaceId] + " did not match specified interface id " + interfId)
            }
            return DBKeyPair(interfaceId: result[0][interfaceId], publicKey: result[0][publicKey], privateKey: result[0][privateKey])
        } else {
            return nil
        }
    }
    
    /**
     Persistently stores a public key associated with an interface ID.
     
     - Parameters:
        - interfaceId: The interface ID of the key pair as a byte array encoded to a hexadecimal `String`.
        - publicKey: The public key to be stored, as a byte array encoded to a hexadecimal `String`.
     
     - Throws: `DBServiceError.unexpectedDbResult` if a public key with the given interface ID is already found in the database.
     */
    func storePublicKey(interfaceId interf_id: String, publicKey pub_key: String) throws {
        guard try getPublicKey(interfaceId: interf_id) == nil else {
            throw DBServiceError.unexpectedDbResult(message: "Database query for public key with interface id " + interf_id + " returned result")
        }
        try db!.run(publicKeys.insert(interfaceId <- interf_id, publicKey <- pub_key))
    }
    
    /**
     Retrieves the persistently stored public key associated with the given interface ID, or returns `nil` if no such key is found.
     
     - Parameter interfaceId: The interface ID of the public key as a byte array encoded to a hexadecimal `String`.
     
     - Returns: A `DBPublicKey` if a public key with the specified interface ID is found, or `nil` if no such public key is found.
     
     - Throws: `DBServiceError.unexpectedDbResult` if multiple public keys are found for a given interface ID, or if the interface ID of the public key returned from the database query does not match `interfaceId`.
     */
    func getPublicKey(interfaceId interfId: String) throws -> DBPublicKey? {
        let rowIterator = try db!.prepareRowIterator(publicKeys.filter(interfaceId == interfId))
        let result = try Array(rowIterator)
        if (result.count > 1) {
            throw DBServiceError.unexpectedDbResult(message: "Multiple public keys found with given interface id " + interfId)
        } else if (result.count == 1) {
            guard result[0][interfaceId] == interfId else {
                throw DBServiceError.unexpectedDbResult(message: "Returned interface id " + result[0][interfaceId] + " did not match specified interface id " + interfId)
            }
            return DBPublicKey(interfaceId: result[0][interfaceId], publicKey: result[0][publicKey])
        } else {
            return nil
        }
    }
    
}
